### DISCUSSION TITLE
Nice Projekt but where are the Server docs

### DISCUSSION THREAD

**Discussion by Not-Zero-Blank** [vpnhood/VpnHood#80] [Q&A]
So i spend now like 8hours to understand the Whole Server and i hope im not just Blind and didnt find the Docs.
Anyway my Question is in the Project File "VpnHood.Server.Access" is the cs RestAccessServer in line 131 its creating there a API request and im wondering if i can use my RestAPI to authenticate my Friends that only my Friends can Access the VPN except that if i understand it right i can turn on the Maintenance Mode with Sending a Connection Refused from my API.
Thanks for every Answer i hope im not just dumb.
![Uploading Bild_2021-12-19_154225.png…]()

**Comment by trudyhood** [comment]
Dear Yeet, There is no source code document, as the project is in the early stage and we update it frequently.
Yes, you can use RestAPI to authenticate your users, but the problem is that how you are going to authenticate them on the client-side? We haven't implemented any login yet, but the best way is just to give them a separate access token. Access Tokens are like keys, only people with access tokens can connect to your server.

We have also launched an experimental portal for VpnHood, please check it here:
https://github.com/vpnhood/VpnHood/wiki/VpnHood-Access-Server

Sorry that there is no documentation for code, but recently software developers don't create many documents for their codes, let me know if you have any questions.

**Reply by Not-Zero-Blank** [reply]
So to your Question how i authenticate my users on the client side i have a rest API linked with a DB. I was just asking because it will took work from me creating my own Auth System. Respect for this Project its very Nice i love it :)

**Reply by Not-Zero-Blank** [reply]
Ah and my question is is the base Url the Url from the token or do i need to set in anywhere?

**Reply by Not-Zero-Blank** [reply]
I understood now.
RestAccessServer is to Setup the API Request before connecting.

**Reply by trudyhood** [reply]
Yes, if you are going to use RestAccessServer, you need to set it up first. Make sure you keep the server up-to-date. It has some known issues and subject to change frequently. Unfortunately, we can't support the legacy versions.

Feel free to ask me any more questions you have.

**Reply by Not-Zero-Blank** [reply]
Im stuck at getting the SSL Certification to work for the RestAccessServer 
Exception:
System.IO.IOException: The decryption operation failed, see inner exception.
 ---> Interop+OpenSsl+SslException: Decrypt failed with OpenSSL error - SSL_ERROR_SSL.
 ---> Interop+Crypto+OpenSslCryptographicException: error:14094416:SSL routines:ssl3_read_bytes:sslv3 alert certificate unknown
   --- End of inner exception stack trace ---
   at Interop.OpenSsl.Decrypt(SafeSslHandle context, Span`1 buffer, SslErrorCode& errorCode)
   at System.Net.Security.SslStreamPal.DecryptMessage(SafeDeleteSslContext securityContext, Span`1 buffer, Int32& offset, Int32& count)
   --- End of inner exception stack trace ---

**Reply by trudyhood** [reply]
Do you use valid certificates or self signed?

**Reply by Not-Zero-Blank** [reply]
> 

Self Signed i use the code from the FileAccessServer shoud i use a Cloudflare cert?

**Reply by trudyhood** [reply]
You have many ways to overcome with it:

**Method 1:** Install a valid certificate on your access server, check free SSL 
**Method 2:** Use services like Cloudflare for your access server as you said
**Method 3:** Use self-certificate for your server, but install the certificate as the root trusted certificate in your VpnHood Server too, in this case, it would be valid for your VpnHoodServer Machine
**Method 4:** Set your certificate thumbprint in CertificateThumbprint field in appsettings.json so VpnHoodServer accept your self-signed certificate like this:

```json
"RestAccessServer": {
    "BaseUrl": "https://....",
    "Authorization": "Bearer ",
   "CertificateThumbprint": "<put your serlf-certificate thumbprint here>"
  }
```

I hope this help you.

**Reply by Not-Zero-Blank** [reply]
its a while ago now... anyway i installed a root trusted certificate from cloudflare but still doesnt work still the same error...

**Reply by trudyhood** [reply]
Would you let me know how did you install the cert on your VpnHood Server? It would be enough to just create a pfx from your cert and replace default.pfx
